pkg/beacon/relay/event/deduplicator.go

```diff
@@ -0,0 +1,132 @@
+type Deduplicator struct {
+	chain                           chain
+	minGroupSelectionDurationBlocks uint64
+
+	groupSelectionMutex             sync.Mutex
+	currentGroupSelectionStartBlock uint64
+
+	relayEntryMutex             sync.Mutex
+	currentRequestStartBlock    uint64
+	currentRequestPreviousEntry string
+}
+
+// NewDeduplicator constructs a new Deduplicator instance.
+func NewDeduplicator(
+	chain chain,
+	minGroupSelectionDurationBlocks uint64,
+) *Deduplicator {
+	return &Deduplicator{
+		chain:                           chain,
+		minGroupSelectionDurationBlocks: minGroupSelectionDurationBlocks,
+	}
+}
+
+// NotifyGroupSelectionStarted notifies the client wants to start group
+// selection upon receiving an event. It returns boolean indicating whether the
+// client should proceed with the execution or ignore the event as a duplicate.
+func (d *Deduplicator) NotifyGroupSelectionStarted(
+	newGroupSelectionStartBlock uint64,
+) bool {
+	d.groupSelectionMutex.Lock()
+	defer d.groupSelectionMutex.Unlock()
+
+	minCurrentGroupSelectionEndBlock := d.currentGroupSelectionStartBlock +
+		d.minGroupSelectionDurationBlocks
+
+	shouldUpdate := d.currentGroupSelectionStartBlock == 0 ||
+		newGroupSelectionStartBlock > minCurrentGroupSelectionEndBlock
+
+	if shouldUpdate {
+		d.currentGroupSelectionStartBlock = newGroupSelectionStartBlock
+		return true
+	}
+
+	return false
+}
+
+// NotifyRelayEntryStarted notifies the client wants to start relay entry
+// generation upon receiving an event. It returns boolean indicating whether the
+// client should proceed with the execution or ignore the event as a duplicate.
+func (d *Deduplicator) NotifyRelayEntryStarted(
+	newRequestStartBlock uint64,
+	newRequestPreviousEntry string,
+) (bool, error) {
+	d.relayEntryMutex.Lock()
+	defer d.relayEntryMutex.Unlock()
+
+	shouldUpdate := func() (bool, error) {
+		// If there is no prior relay request registered by this node, return
+		// true unconditionally.
+		if d.currentRequestStartBlock == 0 {
+			return true, nil
+		}
+
+		// A valid new relay request should have its block number bigger than
+		// the current one because it occurs later for sure.
+		if newRequestStartBlock > d.currentRequestStartBlock {
+			// There may be a case when new relay request holds the same
+			// previous entry than the current one. It is the case when a timed
+			// out request is retried. In that case, we must verify the chain
+			// state. In contrary, if new relay request holds a different
+			// previous entry than the current one, everything is ok.
+			if newRequestPreviousEntry == d.currentRequestPreviousEntry {
+				currentRequestPreviousEntryOnChain, err := d.chain.
+					CurrentRequestPreviousEntry()
+				if err != nil {
+					return false, fmt.Errorf(
+						"could not get current request previous entry: [%v]",
+						err,
+					)
+				}
+
+				if newRequestPreviousEntry ==
+					hex.EncodeToString(currentRequestPreviousEntryOnChain[:]) {
+					return true, nil
```

Comment:
I think we cannot escape taking into account the time it takes to calculate relay entry off-chain.

Scenarios we observed on mainnet with no deduplication:
- The same previous entry but block number higher by 1 or 2 (minor chain reorg)
- The same previous entry and block number significantly higher (another group retrying)
- Another previous entry and block number significantly higher (happy path)

Scenarios we will observe on mainnet with deduplication:
- The same previous entry but block number higher by 1 or 2 (minor chain reorg)
- The same previous entry and block number significantly higher (another group retrying)
- Another previous entry and block number significantly higher (happy path)
- The same previous entry, block number from the past (event pull)
- Another previous entry, block number from the past (event pull, multiple requests)

Still possible, but the probability is quite low:
- Two requests with the same block number but different previous entries (chain fork)
- Two requests with block number close to each other but different previous entries (chain fork)

With the current implementation, the first scenario described (the same previous entry but block number higher by 1 or 2 (minor chain reorg)) will return true although the event should be filtered out.

I assume that in all cases but the old event pull, the new event received will have the block number higher or equal to the last one. This includes reorgs - if Ethereum client switches to another fork, that fork is longer. Taking it into account, I'd propose the following algorithm:
- If the previous entry is different and the block is higher than the previous one - let it pass.
- If the previous entry is the same and the block is higher than the previous one at least for the number of blocks it takes to sign entry off-chain - let it pass.
- Otherwise, filter out as a duplicate.

Reply:
FWIW, I wouldn't care about reorgs for DKG. The penalty is not that severe and it takes a long time to create a group off-chain.

Reply:
You're right about the missing pieces about minor reorgs. However, we can't determine the time needed to calculate relay entry off-chain because the relay entry protocol is time-based and doesn't use the state machine (I mentioned this in one of the above comments). Anyway, I think we should just additionally check the current request starting block on-chain. This should protect us against duplicated events coming with a slightly higher block number than the original one. I added this here: https://github.com/keep-network/keep-core/pull/2574/commits/951a2bd578b0d6b7b3fd84ae4deb11a66a841e0d. What do you think?